src/platform/ESP32/ESP32FactoryDataProvider.cpp

```diff
@@ -154,5 +154,75 @@ CHIP_ERROR ESP32FactoryDataProvider::SignWithDeviceAttestationKey(const ByteSpan
     return CopySpanToMutableSpan(ByteSpan{ signature.ConstBytes(), signature.Length() }, outSignBuffer);
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_DEVICE_INSTANCE_INFO_PROVIDER
+CHIP_ERROR ESP32FactoryDataProvider::GetVendorName(char * buf, size_t bufSize)
+{
+    ChipError err        = CHIP_NO_ERROR;
+    size_t vendorNameLen = 0; // without counting null-terminator
+
+    err = ESP32Config::ReadConfigValueStr(ESP32Config::kConfigKey_VendorName, buf, bufSize, vendorNameLen);
+    ReturnErrorOnFailure(err);
+
+    ReturnErrorCodeIf(vendorNameLen >= bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(buf[vendorNameLen] != 0, CHIP_ERROR_INVALID_STRING_LENGTH);
+
+    return err;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetVendorId(uint16_t & vendorId)
+{
+    ChipError err   = CHIP_NO_ERROR;
+    uint32_t valInt = 0;
+
+    err = ESP32Config::ReadConfigValue(ESP32Config::kConfigKey_VendorId, valInt);
+    ReturnErrorOnFailure(err);
+
+    vendorId = static_cast<uint16_t>(valInt);
+
+    return err;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetProductName(char * buf, size_t bufSize)
+{
+    ChipError err         = CHIP_NO_ERROR;
+    size_t productNameLen = 0; // without counting null-terminator
+
+    err = ESP32Config::ReadConfigValueStr(ESP32Config::kConfigKey_ProductName, buf, bufSize, productNameLen);
+    ReturnErrorOnFailure(err);
+
+    ReturnErrorCodeIf(productNameLen >= bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
+    ReturnErrorCodeIf(buf[productNameLen] != 0, CHIP_ERROR_INVALID_STRING_LENGTH);
+
+    return err;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetProductId(uint16_t & productId)
+{
+    ChipError err   = CHIP_NO_ERROR;
+    uint32_t valInt = 0;
+
+    err = ESP32Config::ReadConfigValue(ESP32Config::kConfigKey_ProductId, valInt);
+    ReturnErrorOnFailure(err);
+
+    productId = static_cast<uint16_t>(valInt);
+
+    return err;
+}
+
+CHIP_ERROR ESP32FactoryDataProvider::GetHardwareVersionString(char * buf, size_t bufSize)
+{
+    ChipError err                   = CHIP_NO_ERROR;
+    size_t hardwareVersionStringLen = 0; // without counting null-terminator
+
+    err = ESP32Config::ReadConfigValueStr(ESP32Config::kConfigKey_HardwareVersionString, buf, bufSize, hardwareVersionStringLen);
+    ReturnErrorOnFailure(err);
+
+    ReturnErrorCodeIf(hardwareVersionStringLen >= bufSize, CHIP_ERROR_BUFFER_TOO_SMALL);
```

Comment:
It really feels like these various methods that all read a string but with different keys should be one helper method that takes key as argument, so we don't copy/paste the same code multiple times...

Reply:
So, `ESP32Config::ReadConfigValueStr` has a check for `CHIP_ERROR_BUFFER_TOO_SMALL` and I moved the `CHIP_ERROR_INVALID_STRING_LENGTH ` check to the same function. Please take a look.